src/theory/quantifiers/quantifiers_rewriter.cpp

```diff
@@ -300,6 +281,57 @@ RewriteResponse QuantifiersRewriter::postRewrite(TNode in)
   return RewriteResponse( status, ret );
 }
 
+Node QuantifiersRewriter::mergePrenex(const Node& q)
+{
+  Assert(q.getKind() == Kind::FORALL || q.getKind() == Kind::EXISTS);
+  Kind k = q.getKind();
+  std::vector<Node> boundVars;
+  Node body = q;
+  bool combineQuantifiers = false;
+  bool continueCombine = false;
+  do
+  {
+    for (const Node& v : body[0])
+    {
+      if (std::find(boundVars.begin(), boundVars.end(), v) == boundVars.end())
+      {
+        boundVars.push_back(v);
+      }
+    }
+    continueCombine = false;
+    if (body.getNumChildren() == 2 && body[1].getKind() == k)
+    {
+      QAttributes qa;
+      QuantAttributes::computeQuantAttributes(body[1], qa);
+      // should never combine a quantified formula with a pool or
+      // non-standard quantified formula here.
+      // note that we technically should check
+      // doOperation(body[1], COMPUTE_PRENEX, qa) here, although this
+      // is too restrictive, as sometimes nested patterns should just be
+      // applied to the top level.
```

Comment:
This comment could be more clear. I assume you mean that you'd be able to merge more if you first prenexed the body, right? Moreover, the justification for why it is too restrictive is not clear to me.  Please clarify and give an example.

Reply:
FYI this code is not new and is just moved in this PR but I can revisit this.

Reply:
I noticed but I took the opportunity. :)